### PR TITLE
Add forex support

### PR DESCRIPTION
**Pull request by harshrathore-msft**
The aim of this PR is:
1. To add support for multiple currencies - Since, ASR and Backup costs are not obtained from Azure Migrate APIs, these costs need to be calculated by AME.
2. Add support for price conversion with business case - Business case only supports USD so for holistic price comparison, this needs to be converted to provided currencies.

### CODE REVIEW

File: src/Models/JSONResponses/Forex/ForexJSON.cs [hunk 1]
```diff
@@ -0,0 +1,20 @@
+﻿using Newtonsoft.Json;
+using System.Collections.Generic;
+
+namespace Azure.Migrate.Export.Models
+{
+    public class ForexJSON
```
Comment (ronehr-microsoft) [inline src/Models/JSONResponses/Forex/ForexJSON.cs:6]: nit. ForexData

File: src/Exchange_Rates.json [hunk 2]
```diff
@@ -0,0 +1,23 @@
+{
+  "success": true,
+  "base": "USD",
```
Comment (ronehr-microsoft) [inline src/Exchange_Rates.json:3]: baseCurrency

File: src/Exchange_Rates.json [hunk 3]
```diff
@@ -0,0 +1,23 @@
+{
+  "success": true,
```
Comment (ronehr-microsoft) [inline src/Exchange_Rates.json:2]: Why do we need this?

File: src/Exchange_Rates.json [hunk 4]
```diff
@@ -0,0 +1,23 @@
+{
+  "success": true,
+  "base": "USD",
+  "date": "2023-07-17",
```
Comment (ronehr-microsoft) [inline src/Exchange_Rates.json:4]: ratesTimestamp

File: src/Exchange_Rates.json (hunk 3 above)
Comment (harshrathore-msft) [reply src/Exchange_Rates.json:2]: We don't need it and it is not really used anywhere as well. It is just another form of 200 OK status returned by the API as JSON, **I can remove this**. Thanks.

File: src/Exchange_Rates.json (hunk 2 above)
Comment (harshrathore-msft) [reply src/Exchange_Rates.json:3]: Hello @ronehr-microsoft, this is the format as returned by the Forex API in its JSON response. 

So I am using the same JSON model for the cache file as well as, for the JSON response from API.

Should I create 2 different JSON models one for the API response and one for the cached file?

File: src/Exchange_Rates.json (hunk 4 above)
Comment (harshrathore-msft) [reply src/Exchange_Rates.json:4]: Hello @ronehr-microsoft, this is the format as returned by the Forex API in its JSON response. 

So I am using the same JSON model for the cache file as well as, for the JSON response from API.

Should I create 2 different JSON models one for the API response and one for the cached file?

File: src/Exchange_Rates.json (hunk 3 above)
Comment (ronehr-microsoft) [reply src/Exchange_Rates.json:2]: Yes, let's remove anything unnecessary.

File: src/Exchange_Rates.json (hunk 2 above)
Comment (ronehr-microsoft) [reply src/Exchange_Rates.json:3]: That's fine. Let's keep the model consistent, adding a transformer increases complexity.

File: src/Common/Enum.cs [hunk 5]
```diff
@@ -144,6 +144,24 @@ public enum IssueCategories
         Internal
     }
 
+    public enum ExchangeRateStates
```
Comment (ronehr-microsoft) [inline src/Common/Enum.cs:147]: Should we rename the enum to "ExchangeRateSource" and keep only 3 values: USDOnly, File, Memory.
Default value = USDOnly. 
Call exchange api to fetch rates. If successful - update enum to Memory.
Read from json to fetch rates. If successful - update enum to File.

File: src/Exchange_Rates.json [hunk 6]
```diff
@@ -0,0 +1,23 @@
+{
+  "success": true,
+  "base": "USD",
+  "date": "2023-07-17",
+  "rates": {
```
Comment (ronehr-microsoft) [inline src/Exchange_Rates.json:5]: Are these all the currencies that we support in AME? If yes, then, we don't need a USD fallback.

File: src/Forex/ForexData.cs [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+        {
+            return ExchangeRateState;
+        }
+
+        public static double GetExchangeRate()
+        {
+            return ExchangeRate;
+        }
+
+        public static void UpdateExchangeRatesWithUSDFallback()
```
Comment (ronehr-microsoft) [inline src/Forex/ForexData.cs:33]: Do we have exchange rates for all currencies that we support in AME? If yes, then, we don't need a USD fallback.

File: src/Forex/ForexData.cs [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+        {
+            List<KeyValuePair<string, string>> currencies = InitializationData.GetSupportedCurrenciesInitializationData();
+            ExchangeRatesUSD = new Dictionary<string, double>();
+
+            foreach(var kvp in currencies)
+                if (!ExchangeRatesUSD.ContainsKey(kvp.Key))
+                    ExchangeRatesUSD.Add(kvp.Key, 1.0);
```
Comment (ronehr-microsoft) [inline src/Forex/ForexData.cs:40]: This leads to incorrect data used as workaround. Let's avoid this.

File: src/Common/Enum.cs (hunk 5 above)
Comment (harshrathore-msft) [reply src/Common/Enum.cs]: Removed the concept of ExchangeRateStates completely, state is determined whether ExchangeRates dictionary is null or not.

File: src/Exchange_Rates.json (hunk 6 above)
Comment (harshrathore-msft) [reply src/Exchange_Rates.json:4]: Removed USD fallback

File: src/Forex/ForexData.cs (hunk 7 above)
Comment (harshrathore-msft) [reply src/Forex/ForexData.cs]: Thanks for the great suggestion, removed USD fallback

File: src/Forex/ForexData.cs (hunk 8 above)
Comment (harshrathore-msft) [reply src/Forex/ForexData.cs]: Yup, thanks for the great suggestion, removed the population of false data in dictionary.

File: src/Assessment/Assess.cs [hunk 9]
```diff
@@ -566,6 +569,26 @@ public bool BeginAssessment()
...
+        {
+            UserInputObj.LoggerObj.LogInformation("Obtaining exchange rates' data");
+            double exchangeRate = 1.0;
+            try
+            {
+                exchangeRate = ForexData.GetExchangeRate(UserInputObj);
+            }
+            catch (OperationCanceledException)
+            {
+                throw;
```
Comment (ronehr-microsoft) [inline src/Assessment/Assess.cs:582]: We should not throw. Maybe just log warning and keep using USD

File: src/Assessment/Assess.cs [hunk 10]
```diff
@@ -566,6 +569,26 @@ public bool BeginAssessment()
             return true;
         }
 
+        private void UpdateExchangeRates()
+        {
+            UserInputObj.LoggerObj.LogInformation("Obtaining exchange rates' data");
+            double exchangeRate = 1.0;
```
Comment (ronehr-microsoft) [inline src/Assessment/Assess.cs:575]: How should we communicate back to the user that we are using USD incase of failures? Should we update the excel output with currency = USD along with warning message below?

File: src/Forex/ForexData.cs [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+using Azure.Migrate.Export.HttpRequestHelper;
+using Azure.Migrate.Export.Models;
+
+namespace Azure.Migrate.Export.Forex
+{
+    public class ForexData
+    {
+        private static Dictionary<string, double> ExchangeRatesUSD = null;
+
+        public static double GetExchangeRate(UserInput userInputObj)
```
Comment (ronehr-microsoft) [inline src/Forex/ForexData.cs:16]: Here we should just pass currency, not the whole userInputObj.

File: src/Forex/ForexData.cs [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+                ExchangeRatesUSD.Count <= 0 ||
+                !ExchangeRatesUSD.ContainsKey(userInputObj.Currency.Key))
+            {
+                return 1.0;
+            }
+
+            return ExchangeRatesUSD[userInputObj.Currency.Key];
+        }
+
+        private static void GetExchangeRatesFromFile(UserInput userInputObj)
```
Comment (ronehr-microsoft) [inline src/Forex/ForexData.cs:30]: Not needed.

File: src/Forex/ForexData.cs [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+            if (difference >= 30.0)
+            {
+                userInputObj.LoggerObj.LogWarning("Cached exchange rates are more than 30 days old");
+                return;
+            }
+
+            ExchangeRatesUSD = forexJSONObj.Rates;
+        }
+
+        private static void GetExchangeRatesFromAPI(UserInput userInputObj)
```
Comment (ronehr-microsoft) [inline src/Forex/ForexData.cs:57]: Not needed to pass input param here

File: src/Assessment/Assess.cs (hunk 9 above)
Comment (harshrathore-msft) [reply src/Assessment/Assess.cs:582]: We only throw on "OperationCanceledException" -> This exception is caused when the user tries to "Stop" the process from the UI.

When the stop button is clicked, it sets a cancellationcontext. This cancellation context is read by all threads and periodically checked at a stable point during their execution. Once the cancellation context is set, OperationCancelled exception is thrown.

File: src/Forex/ForexData.cs (hunk 11 above)
Comment (harshrathore-msft) [reply src/Forex/ForexData.cs:16]: There are three reasons for passing UserInput parameter here:
1. Firstly is the CurrencySymbol (USD, INR, etc)
2. The logger object is also a part of this parameter, hence to record warning / errors during API calls require it for updating the logs.
3. If everything fails, and we are not able to obtain any exchange rate -> we need to reset UserInputObject.Currency to USD.

Comment (harshrathore-msft) [reply src/Forex/ForexData.cs]: Understood, Implemented ForexData with singleton parameter.
Thanks.

File: src/Forex/ForexData.cs (hunk 12 above)
Comment (harshrathore-msft) [reply src/Forex/ForexData.cs]: Understood, Implemented ForexData with singleton parameter.
Thanks.

File: src/Forex/ForexData.cs (hunk 13 above)
Comment (harshrathore-msft) [reply src/Forex/ForexData.cs]: Understood, Implemented ForexData with singleton parameter.
Thanks.

File: src/Forex/ForexData.cs [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+            ExchangeRatesUSD = null;
+        }
+
+        private ForexData()
+        {
+            UserInputObj = new UserInput();
+            ExchangeRatesUSD = null;
+        }
+
+        public static ForexData GetInstance(UserInput userInputObj)
```
Comment (ronehr-microsoft) [inline src/Forex/ForexData.cs:36]: There should be single contructor and GetInstance() as its a singleton class. Only one object is created for the class.

**Comment by harshrathore-msft** [comment]
Closing due to change in design for FOREX

**Pull request closed without merging**